core/driver/sources/src/main/scala/org/virtuslab/ideprobe/config/IntelliJProvider.scala

```diff
@@ -0,0 +1,96 @@
+package org.virtuslab.ideprobe.config
+
+import com.google.gson.Gson
+
+import java.nio.file.{CopyOption, FileVisitResult, Files, Path, SimpleFileVisitor}
+import org.virtuslab.ideprobe.ide.intellij.{InstalledIntelliJ, IntelliJFactory, LocalIntelliJ}
+import org.virtuslab.ideprobe.IdeProbePaths
+import org.virtuslab.ideprobe.dependencies.{IntelliJVersion, Plugin}
+
+import java.nio.file.attribute.BasicFileAttributes
+import scala.io.Source
+
+trait IntelliJProvider {
+  def factory: IntelliJFactory
+  def setup(): InstalledIntelliJ
+  def version: IntelliJVersion
+  def withFactory(factory: IntelliJFactory): IntelliJProvider
+}
+
+case class ExistingIntelliJ(
+    path: Path,
+    plugins: Seq[Plugin],
+    factory: IntelliJFactory
+) extends IntelliJProvider {
+  override def setup(): InstalledIntelliJ = {
+    val pluginsDir = path.resolve("plugins")
+    val backupDir = Files.createTempDirectory(path, "plugins")
+    copyFolder(pluginsDir, backupDir)
+    factory.installPlugins(plugins, path)
+
+    new LocalIntelliJ(path, factory.paths, factory.config, backupDir)
+  }
+
+  private def copyFolder(source: Path, target: Path): Unit = {
+    Files.walkFileTree(
+      source,
+      new SimpleFileVisitor[Path] {
+
+        override def preVisitDirectory(dir: Path, attrs: BasicFileAttributes): FileVisitResult = {
+          Files.createDirectories(target.resolve(source.relativize(dir)))
+          FileVisitResult.CONTINUE
+        }
+
+        override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
+          Files.copy(file, target.resolve(source.relativize(file)))
+          FileVisitResult.CONTINUE
+        }
+      }
+    )
+  }
+
+  override lazy val version: IntelliJVersion = {
+    case class ProductInfo(version: String, buildNumber: String)
+    val productInfo = path.resolve("product-info.json").toFile
+    val source = Source.fromFile(productInfo)
+    val content = source.mkString
+    source.close
+    val gson = new Gson
+    val ProductInfo(version, buildNumber) = gson.fromJson(content, classOf[ProductInfo])
+
+    IntelliJVersion(version, Some(buildNumber))
+  }
+
+  override def withFactory(factory: IntelliJFactory): IntelliJProvider = copy(factory = factory)
+}
+
+case class DefaultIntelliJ(
+    version: IntelliJVersion,
+    plugins: Seq[Plugin],
+    factory: IntelliJFactory
+) extends IntelliJProvider {
+
+  override def setup(): InstalledIntelliJ =
+    factory.create(version, plugins)
```

Comment:
I would refactor here. `ExistingIntelliJ` in `setup` contains the setup logic but the setup logic of `DefaultIntelliJ` is located in factory. Any idea if there is a better approach?

Reply:
The two setup processes are fundamentally different - handling existing instances is outside of factory's responsibilities (at least in its current form). The provider makes a decision which approach should be taken, and depending on that either delegates the creation to factory, or just prepares the pre-installed instance for further manipulation.

We could have two specialized "factories" for both default and existing instances, thus eliminating the need for a provider completely. @lukaszwawrzyk WDYT?

Reply:
This is kind of what I meant. IJFactory should be renamed to IJProvider and the other "factory"/provider should handle preparing existing intellij

Reply:
Okay, will be done in the next commit.

Reply:
Done.